### PR TITLE
[pack] Ensuring JobHost restart suppresion aligns with request lifecycle

### LINKED ISSUE #10631: Azure Functions Host 1036 results in a 500 Error on a PUT to /admin/functions
#### Check for a solution in the Azure portal
For issues in production, please check for a solution to common issues in the Azure portal before opening a bug. In the Azure portal, navigate to your function app, select `Diagnose and solve problems` from the left, and view relevant dashboards before opening your issue.

#### Investigative information

Please provide the following:

- Timestamp: 2024-11-13 22:13:56.4826033 to 2024-11-13 22:19:42.4812748
- Function App version: 4.1036.3.23284
- Function App name: testfinnodeFGG
- Function name(s) (as appropriate):HttpTrigger1
- Invocation ID:
- Region: East US

#### Repro steps

Provide the steps required to reproduce the problem:

1) Create a new Function App targeting Node 18 or 20 to allow Portal Create.
2) Try and add a Function via the Portal and see that it fails with 500 Server error. 
3) The Function does get created but the Script Host restarts. 
4) Pin back to 4.36.0 and try step 2 again. There are no errors. 

#### Expected behavior

Provide a description of the expected behavior.

1) The Function create should not result in a 500 Server error.

#### Actual behavior

Provide a description of the actual behavior observed.

1) The Portal Function Function creation calls /admin/functions and this call returrns a 500 error but the files do get created.
2) The Host detects a file change notification from the Put and Initiates a restart.
3) There is an exception during the restart with this call stack.

System.ObjectDisposedException : Cannot access a disposed object. 
Object name: 'IServiceProvider'.
at Microsoft.Extensions.DependencyInjection.ServiceLookup.ThrowHelper.ThrowObjectDisposedException()
at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier,ServiceProviderEngineScope serviceProviderEngineScope)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider,Type serviceType)
at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
at Microsoft.AspNetCore.Mvc.ObjectResult.ExecuteResultAsync(ActionContext context)
at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultAsync(IActionResult result)
at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next,Scope& scope,Object& state,Boolean& isCompleted)

The ARM Template is from DataDog and an example deployment creates the Files in the ARM Template e.g [datadog-serverless-functions/azure/deploy-to-azure/function_template.json at master · DataDog/datadog-serverless-functions](https://github.com/DataDog/datadog-serverless-functions/blob/master/azure/deploy-to-azure/function_template.json)

#### Known workarounds

Provide a description of any known workarounds.

Pin back to 4.36

#### Related information 

Provide any related information 

* Programming language used: Node
* Links to source : https://github.com/DataDog/datadog-serverless-functions/tree/master/azure/activity_logs_monitoring
* Bindings used : HTTP

### PR DESCRIPTION
**Pull request by fabiocav**
### Issue describing the changes in this PR

resolves #10631

This change aligns the restart suspension scope with the request lifecycle, ensuring that a JobHost restart triggered by changes to the functions does not resume before the request is complete, which could otherwise result in its services being disposed prematurely.

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 1]
```diff
@@ -80,12 +81,10 @@ public async Task<IActionResult> CreateOrUpdate(string name, [FromBody] Function
...
             }
 
-            bool success, configChanged;
-            FunctionMetadataResponse functionMetadataResponse;
-            using (fileMonitoringService.SuspendRestart(true))
-            {
-                (success, configChanged, functionMetadataResponse) = await _functionsManager.CreateOrUpdate(name, functionMetadata, Request);
-            }
+            var restartSuspensionScope = fileMonitoringService.SuspendRestart(true);
+            Response.RegisterForDispose(restartSuspensionScope);
```
Comment (jviau) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs:85]: TIL this exists.

**Review by jviau — APPROVED** [review]

**Comment by kshyju** [comment]
Do we know what introduced this issue? Can we add that details to the PR description to help future readers?

**Comment by kshyju** [comment]
> Do we know what introduced this issue? Can we add that details to the PR description to help future readers?

Noticed in-proc backport not required. So, the DI Changes triggered this?

**Review by kshyju — APPROVED** [review]

**Comment by fabiocav** [comment]
> Do we know what introduced this issue? Can we add that details to the PR description to help future readers?

The pattern previously in place was fragile, and created an opportunity for services to be consumed after disposal. The flow of events was:

1. API was called
2. Restart suppressed
3. Files written (normally, this would trigger a JobHost restart, the suppression prevented that from happening)
4. Restart resumed, processing pending changes
5. API call returned

When step 4 takes place, the JobHost will be terminated, and the JobHost scoped container will be disposed, causing all associated services to be disposed as well. In in-proc, although there is some evidence of the problem happening, so a backport should probably be considered, the restart didn't occur fast enough (or there were no additional service requests) to impact the response handling, with the changes in the default host, the JobHost disposal is taking place before the response is handled and no longer needs to consume services. 

This was a race that was in place from the time that cod was introduced, just not something that was exposed with the previous host.

Hope the above answers the second question as well.

**Comment by kshyju** [comment]
Did you miss the release notes file?

**Review by liliankasem — APPROVED** [review]

**Pull request merged**